File: excelium-cli/src/main/java/excelium/cli/controller/DataSourceController.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.cli.controller;

import excelium.cli.annotation.Command;
import excelium.cli.annotation.Controller;
import excelium.cli.annotation.Injectable;
import excelium.core.database.DatabaseService;
import excelium.core.database.DatabaseServiceFactory;
import excelium.model.enums.DataSourceType;
import excelium.model.project.DataSource;
import excelium.model.project.Project;
import org.apache.commons.lang3.StringUtils;

import javax.xml.bind.JAXBException;
import java.io.IOException;

import static excelium.common.Prompt.*;

/**
 * Provides commands for controlling data source.
 *
 * @author PhungDucKien
 * @since 2018.05.25
 */
@Controller(name = "ds")
public class DataSourceController extends BaseController {

    /**
     * Project instance
     */
    @Injectable
    private Project project;

    /**
     * Adds data source.
     *
     * @throws IOException            the io exception
     * @throws JAXBException          the jaxb exception
     * @throws IllegalAccessException the illegal access exception
     */
    @Command(name = "add")
    public void add() throws IOException, JAXBException {

        String dataSourceTypeId = promptList("What is the type of database you want to connect?", DataSourceType.getListChoice());
        DataSourceType dataSourceType = DataSourceType.fromName(dataSourceTypeId);

        DataSource dataSource = new DataSource();
        dataSource.setType(dataSourceType);

        String databaseName = null;

        if (dataSourceType == DataSourceType.MYSQL || dataSourceType == DataSourceType.POSTGRESQL) {
            String driverClass = dataSourceType.getDriverClass();
            dataSource.setDriverClass(driverClass);

            boolean agree = promptConfirm("Do you want to use connection string?");
            if (agree) {
                String defaultUrl = dataSourceType.getUrl("localhost", String.valueOf(dataSourceType.getDefaultPort()), dataSourceType.getText());
                String url = promptInput("Database Connection String", defaultUrl);
                dataSource.setUrl(url);
            } else {
                String databaseHost = promptInput("Database Host", "localhost");
                String databasePort = promptInput("Database Port", String.valueOf(dataSourceType.getDefaultPort()));
                databaseName = promptInput("Database Name", dataSourceType.getText());

                String url = dataSourceType.getUrl(databaseHost, databasePort, databaseName);
                dataSource.setUrl(url);
            }

            String userName = promptInput("User Name", null);
            String password = promptInput("Password", null);
            dataSource.setUserName(userName);
            dataSource.setPassword(password);
        } else if (dataSourceType == DataSourceType.DYNAMODB) {
            databaseName = dataSourceType.getText();
        }

        DatabaseService databaseService = DatabaseServiceFactory.createService(dataSource);
        boolean isConnectible = databaseService.testConnection();

        if (isConnectible) {
            String dataSourceName = promptInput("Enter the name of the data source", databaseName);
            dataSource.setName(dataSourceName);
            project.addDataSource(dataSource);

            if (project.getDataSources().size() == 1) {
                project.setDefaultDataSource(dataSource.getName());
            } else {
                String defaultDataSource = promptList("Please choose the default data source", project.getDataSourceListChoice(false));
                project.setDefaultDataSource(defaultDataSource);
            }

            updateProjectFile(project);
        } else {
            System.out.println("The database is not connectible.");
        }
    }

    /**
     * Removes data source.
     *
     * @throws IOException            the io exception
     * @throws JAXBException          the jaxb exception
     * @throws IllegalAccessException the illegal access exception
     */
    @Command(name = "remove")
    public void remove() throws IOException, JAXBException {
        String removeDataSource = promptList("Choose the data source to remove:", project.getDataSourceListChoice(true));
        if (removeDataSource.equals(DataSource.ALL)) {
            project.getDataSources().clear();
            project.setDefaultDataSource(null);
        } else {
            project.getDataSources().remove(removeDataSource);
            if (StringUtils.equals(project.getDefaultDataSource(), removeDataSource)) {
                if (project.getDataSources().size() == 0) {
                    project.setDefaultDataSource(null);
                } else if (project.getDataSources().size() == 1) {
                    project.setDefaultDataSource(project.getDataSources().keySet().iterator().next());
                } else {
                    String defaultDataSource = promptList("Please choose the default data source", project.getDataSourceListChoice(false));
                    project.setDefaultDataSource(defaultDataSource);
                }
            }
        }

        updateProjectFile(project);
    }
}

File: packages/cli/src/controller/DataSourceController.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Prompt } from '@excelium/common';
import { DataSource, DataSourceType, Project } from '@excelium/model';
import { inject } from 'inversify';
import BeanFactory from '../BeanFactory';
import { action, group } from '../commander';
import BaseController from './BaseController';

/**
 * Provides commands for controlling data source.
 *
 * @author PhungDucKien
 * @since 2018.05.25
 */
@group({ name: 'ds' })
export default class DataSourceController extends BaseController {
  /**
   * Project instance
   */
  @inject(BeanFactory.projectProvider)
  protected projectProvider: () => Promise<Project>;

  /**
   * Adds data source.
   */
  @action({ name: 'add', options: BaseController.baseParameters })
  public async add(): Promise<void> {
    const project = await this.projectProvider();

    const dataSourceTypeId = await Prompt.promptList('What is the type of database you want to connect?', DataSourceType.getListChoice());
    const dataSourceType = DataSourceType.fromName(dataSourceTypeId);

    const dataSource = new DataSource();
    dataSource.type = dataSourceType;

    // TODO
    // let databaseName;

    // if (dataSourceType === DataSourceType.MYSQL || dataSourceType === DataSourceType.POSTGRESQL) {
    //   const driverClass = dataSourceType.getDriverClass();

    //   const databaseHost = await Prompt.promptInput('Database Host', 'localhost');
    //   const databasePort = await Prompt.promptInput('Database Port', String.valueOf(dataSourceType.getDefaultPort()));
    //   databaseName = await Prompt.promptInput('Database Name', dataSourceType);
    //   const userName = await Prompt.promptInput('User Name');
    //   const password = await Prompt.promptInput('Password');

    //   const url = dataSourceType.getUrl(databaseHost, databasePort, databaseName);

    //   dataSource.driverClass = driverClass;
    //   dataSource.url = url;
    //   dataSource.userName = userName;
    //   dataSource.password = password;
    // } else if (dataSourceType === DataSourceType.DYNAMODB) {
    //   databaseName = dataSourceType;
    // }

    // const databaseService = DatabaseServiceFactory.createService(dataSource);
    // const isConnectible = databaseService.testConnection();

    // if (isConnectible) {
    //   const dataSourceName = await Prompt.promptInput('Enter the name of the data source', databaseName);
    //   dataSource.name = dataSourceName;
    //   project.addDataSource(dataSource);

    //   if (project.dataSources.size === 1) {
    //     project.defaultDataSource = dataSource.name;
    //   } else {
    //     const defaultDataSource = await Prompt.promptList('Please choose the default data source', project.getDataSourceListChoice(false));
    //     project.defaultDataSource = defaultDataSource;
    //   }

    //   await this.updateProjectFile(project);
    // } else {
    //   console.error('The database is not connectible.');
    // }
  }

  /**
   * Removes data source.
   */
  @action({ name: 'remove', options: BaseController.baseParameters })
  public async remove(): Promise<void> {
    const project = await this.projectProvider();

    const removeDataSource = await Prompt.promptList('Choose the data source to remove:', project.getDataSourceListChoice(true));
    if (removeDataSource === DataSource.ALL) {
      project.dataSources.clear();
      project.defaultDataSource = null;
    } else {
      project.dataSources.delete(removeDataSource);
      if (project.defaultDataSource === removeDataSource) {
        if (project.dataSources.size === 0) {
          project.defaultDataSource = null;
        } else if (project.dataSources.size === 1) {
          project.defaultDataSource = project.dataSources.keys().next().value;
        } else {
          const defaultDataSource = await Prompt.promptList('Please choose the default data source', project.getDataSourceListChoice(false));
          project.defaultDataSource = defaultDataSource;
        }
      }
    }

    await this.updateProjectFile(project);
  }
}
